### PR TITLE
230 satterthwaite

### LINKED ISSUE #230: Satterthwaite degree of freedom for Sandwich Estimator
in`clubSandwich`, there is one implementation for `gls`, see https://github.com/jepusto/clubSandwich/blob/main/R/coef_test.R.
https://github.com/jepusto/clubSandwich/blob/main/R/get_arrays.R
in _Variance estimation for complex surveys using replication techniques_ Rust and Rao suggest using 
image

However, when using similar method for the Satterthwaite d.f. for asymptotic covariance, got very strange result for these sandwich estimatros. The equation is not used, but used another formula

image

- [x] Find out how important is this in applications
- [x] Further check `clubSandwich`
    - [ ] understand the algorithm used in `clubSandwich`
    - [ ] implement the `Satterthwaite` d.f. for the two new added sandwich covariance
    - [ ] or, directly call `clubSandwich` in `mmrm` package (so that the `summary(mmrm_fit)` works)
    - [x] or, implement as the paper describes
- [x] ~~create a vignette that combines all the d.f. things, including residual, Satterthwaite, Kenward-Roger, (between-within), etc; currently they are in different vignettes and not very convenient to understand these things~~ add vignettes in Satterthwaite degree of freedom
- [x] add news

### PR DESCRIPTION
**Pull request by clarkliming**
close #230 

the Satterthwaite degrees of freedom is implemented for sandwich estimator. The scale parameter is not used.

### CODE REVIEW

File: vignettes/satterthwaite.Rmd [hunk 1]
```diff
@@ -119,4 +119,68 @@ of freedom is calculated for each of them. Finally, using properties of the F
 distribution's expectation, the denominator degree of freedom for the whole F statistic
 is derived.
 
+## Satterthwaite degrees of freedom for sandwich covariance
+
+In @bell2002bias the degrees of freedom of sandwich covariance are described.
```
Comment (danielinteractive) [inline vignettes/satterthwaite.Rmd:124]:
```suggestion
In @bell2002bias the Satterthwaite degrees of freedom in combination with a sandwich covariance matrix estimator are described.
```

File: NEWS.md [hunk 2]
```diff
@@ -14,6 +14,8 @@
   for the degrees of freedom, another argument `vcov` is added to specify the
   method used to adjust the coefficients covariance matrix.
 - Add optional `tidymodels` framework support.
+- Add Satterthwaite degrees of freedom support for empirical and empirical Jackknife
```
Comment (danielinteractive) [inline NEWS.md:17]: Maybe merge this with the bullet in line 12 ff.? since empirical cov matrix is only added in the new release

Comment (danielinteractive) [reply NEWS.md:17]: Or leave it separate but then also add the residual d.f. news (I don't see that yet)

File: vignettes/satterthwaite.Rmd [hunk 3]
```diff
@@ -119,4 +119,68 @@ of freedom is calculated for each of them. Finally, using properties of the F
...
+For one-dimensional contrast, following the same notation in [Details of the model fitting in `mmrm`](algorithm.html)
+and [Details of the Kenward-Roger calculations](kenward.html), we have the following derivation.
+For an estimator of variance with the following term
+
+\[
+  v = s c(X^\top X)^{-1}\sum_{i}{X_i^\top A_i \epsilon_i \epsilon_i^\top A_i X_i} (X^\top X)^{-1} c^\top
+\]
+
+where $s$ takes the value of $\frac{n}{n-1}$, $1$ or $\frac{n-1}{n}$, and $A_i$ takes $I_i$, $(I_i - H_{ii})^{-\frac{1}{2}}$, or $(I_i - H_{ii})^{-1}$
+respectively, $c$ is a vector, then $v$ can be decomposed into the a weighted sum of independent $\Chi_1^2$ distribution, where the weights are
```
Comment (danielinteractive) [inline vignettes/satterthwaite.Rmd:137]: the Chi thing does not render locally for me, but maybe it is ok

File: vignettes/satterthwaite.Rmd [hunk 4]
```diff
@@ -119,4 +119,68 @@ of freedom is calculated for each of them. Finally, using properties of the F
...
+\]
+
+$(I - H)_i$ corresponds to the rows of subject $i$.
+
+So the degrees of freedom can be represented as
+\[
+  \nu = \frac{(\sum_{i}\lambda_i)^2}{\sum_{i}{\lambda_i^2}}
+\]
+
+where $\lambda_i$ is the eigen value of matrix $G$.
```
Comment (danielinteractive) [inline vignettes/satterthwaite.Rmd:159]:
```suggestion
where $\lambda_i, i = 1, \dotsc, n$ are the eigenvalues of $G$.
```

File: vignettes/satterthwaite.Rmd [hunk 5]
```diff
@@ -119,4 +119,68 @@ of freedom is calculated for each of them. Finally, using properties of the F
...
+@bell2002bias also suggests that $V$ can be chosen as identify matrix, so $G_{ij} = g_i  g_j$.
```
Comment (danielinteractive) [inline vignettes/satterthwaite.Rmd:160]:
```suggestion
@bell2002bias also suggests that $V$ can be chosen as identify matrix, so $G_{ij} = g_i ^\top g_j$.
```

File: vignettes/satterthwaite.Rmd [hunk 6]
```diff
@@ -119,4 +119,68 @@ of freedom is calculated for each of them. Finally, using properties of the F
...
+
+Following [Weighted Least Square Estimator](algorithm.html#weighted-least-squares-estimator), we can transform
+the original $X$ into $\tilde{x}$ to use the above equations.
+
+To avoid repeated computation of matrix $A_i$, $H$ etc for different contrasts, we calculate and cache the following
+
+\[
+  G^\ast_i = (I - H)_i^\top A_i X_i (X^\top X)^{-1}
+\]
+is a $\sum_i{m_i} \times p$ matrix. With different contrasts, we need only calculate the following
```
Comment (danielinteractive) [inline vignettes/satterthwaite.Rmd:170]:
```suggestion
which is a $\sum_i{m_i} \times p$ matrix. With different contrasts, we need only calculate the following
```

File: vignettes/satterthwaite.Rmd [hunk 7]
```diff
@@ -119,4 +119,68 @@ of freedom is calculated for each of them. Finally, using properties of the F
...
+\[
+  g_i = G^\ast_i c
+\]
+to obtain a $\sum_i{m_i} \times 1$ matrix, $G$ can be computed with $g_i$.
+
+To obtain the degrees of freedom, and to avoid eigen computation on a large matrix, we can use the following equation
```
Comment (danielinteractive) [inline vignettes/satterthwaite.Rmd:176]: is this from the paper? would be good to cite/derive

File: vignettes/satterthwaite.Rmd (hunk 5 above)
Comment (danielinteractive) [reply vignettes/satterthwaite.Rmd:160]: good to check generally on the dimensions also below

**Review by danielinteractive — COMMENTED** [review]
thanks a lot @clarkliming that looks awesome! see few comments inside

File: vignettes/satterthwaite.Rmd [hunk 8]
```diff
@@ -119,4 +119,68 @@ of freedom is calculated for each of them. Finally, using properties of the F
 distribution's expectation, the denominator degree of freedom for the whole F statistic
 is derived.
 
+## Satterthwaite degrees of freedom for sandwich covariance
```
[the remark this answers is not in the document]
Comment (clarkliming) [reply vignettes/satterthwaite.Rmd:122]: The wording seems interchangeable , including sandwich, robust, empirical; here we only implemented the empirical (or, linearization, Huber(-White) sandwich estimator) and Jackknife (CR3) and they have the same form of sandwich estimator. But I agree with you that use the same terminology should be much better

File: vignettes/satterthwaite.Rmd (hunk 7 above)
Comment (clarkliming) [reply vignettes/satterthwaite.Rmd:176]: derivation added

File: vignettes/satterthwaite.Rmd (hunk 8 above)
[the remark this answers is not in the document]
Comment (clarkliming) [reply vignettes/satterthwaite.Rmd]: the reason I use "sandwich" here is that there are different forms of the sandwich estimator, including empirical, BRL, and Jackknife; using the "sandwich" estimator to include these (although empirical and Jackknife included for now) is simpler.

**Review by danielinteractive — APPROVED** [review]
great thanks Liming!

**Pull request merged**